File: services/book-service/src/main/java/com/bookland/book/entity/Book.java
package com.bookland.book.entity;

import lombok.*;
import jakarta.persistence.*;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;
import java.util.stream.Collectors;

@Entity
@Table(name = "book")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Book {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String name;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(nullable = false)
    private Double originalCost;

    @Builder.Default
    private Double sale = 0.0;

    @Builder.Default
    private Integer stock = 0;

    @Enumerated(EnumType.STRING)
    @Builder.Default
    private BookStatus status = BookStatus.ENABLE;

    private LocalDate publishedDate;

    @Column(columnDefinition = "TEXT")
    private String bookImageUrl;

    @Builder.Default
    private Boolean pin = false;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "authorId", nullable = false)
    private Author author;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "publisherId", nullable = false)
    private Publisher publisher;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "seriesId")
    private Serie series;

    // createdBy lưu UUID của user từ identity-service (không FK vì khác DB)
    @Column(name = "created_by")
    private String createdBy;

    @Column(updatable = false)
    private LocalDateTime createdAt;

    private LocalDateTime updatedAt;

    @ManyToMany
    @JoinTable(
            name = "book_category",
            joinColumns = @JoinColumn(name = "bookId"),
            inverseJoinColumns = @JoinColumn(name = "categoryId")
    )
    @Builder.Default
    private Set<Category> categories = new HashSet<>();

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }

    @Transient
    public Double getFinalPrice() {
        if (sale == null || sale == 0) {
            return originalCost;
        }
        return originalCost - (originalCost * sale / 100);
    }

    public enum BookStatus {
        ENABLE, DISABLE
    }
}

File: BookLand_FE/src/types/Book.ts


export const BookStatus = {
    ENABLE: 'ENABLE',
    DISABLE: 'DISABLE'
} as const;

export type BookStatus = (typeof BookStatus)[keyof typeof BookStatus];

export interface Book {
    id: number;
    name: string;
    description?: string;
    originalCost: number;
    sale: number;
    finalPrice: number;
    stock: number;
    status: BookStatus;
    publishedDate?: string;
    bookImageUrl?: string;
    pin: boolean;
    authorId: number;
    authorName: string;
    publisherId: number;
    publisherName: string;
    seriesId?: number;
    seriesName?: string;
    categoryIds: number[];
    rating?: number;
    ratingCount?: number;
    volume?: string | number;
    createdAt?: string;
    updatedAt?: string;
}

// Response từ Elasticsearch /api/books/search — các trường denormalized từ ES index
export interface BookDocument {
    id: string;            // String trong ES (khác với number trong MySQL)
    name: string;
    description?: string;
    originalCost: number;
    sale: number;
    finalPrice: number;
    stock: number;
    status: BookStatus;
    bookImageUrl?: string;
    authorName: string;
    publisherName: string;
    categories: string[];  // Tên thể loại (khác với categoryIds dạng number[])
}

// Hàm normalize BookDocument → Book để tái sử dụng BookGrid và BookCard hiện có
export function normalizeBookDocument(doc: BookDocument): Book {
    return {
        id: Number(doc.id),
        name: doc.name,
        description: doc.description,
        originalCost: doc.originalCost,
        sale: doc.sale,
        finalPrice: doc.finalPrice,
        stock: doc.stock,
        status: doc.status,
        bookImageUrl: doc.bookImageUrl,
        pin: false,
        authorId: 0,
        authorName: doc.authorName,
        publisherId: 0,
        publisherName: doc.publisherName,
        categoryIds: [],
    };
}


export interface BookRequest {
    name: string;
    description?: string;
    originalCost: number;
    sale: number;
    stock: number;
    status: BookStatus;
    publishedDate?: string;
    bookImageUrl?: string;
    pin: boolean;
    authorId: number;
    publisherId: number;
    seriesId?: number;
    categoryIds: number[];
}
